### DISCUSSION TITLE
How to write a Loader that has to decode, decrypt, or decompress the data?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#7186] [Q&A] [answered]
I'm sure this is a very basic question but it's hard to grep the repo to find an example and Googling for Loader dev docs finds a bunch of blog posts from 2020 of varying quality. Is there an official intro?

I've written a bunch of Loaders, all pretty simple. Most are like this:

- `findSupportedLoadSpecs()` - check for magic word if there is one, do sanity check if not, maybe store some offsets for `load()`
- `load()`
  - create a `BinaryReader` from the `ByteProvider`
  - `startAddress = program.getAddressFactory().getDefaultAddressSpace().getAddress(start)`
  - `fileBytes = MemoryBlockUtils.createFileBytes(program, provider, monitor)`
  - `MemoryBlock block = program.getMemory().createInitializedBlock("memory", startAddress, filebytes, offset, length, false)` (false means not an overlay)
  - `entryAddress = program.getAddressFactory().getDefaultAddressSpace().getAddress(entry)`
  - `SymbolTable st = program.getSymbolTable()`
  - `st.createLabel(entryAddress, "entry", SourceType.IMPORTED)`
  - `st.addExternalEntryPoint(entryAddress)`

The `createInitializedBlock` block part directly takes a flat part of the file.

I've got more complex loaders but none that don't take flat parts the file. Which ways do Loader writers use when they have to transform the bytes from the file to get them into a MemoryBlock? Which built-in ones to look at?

**Comment by ryanmkurtz** [comment, marked as answer]
You just use the version of `MemoryBlockUtils.createInitializedBlock()` that takes an `InputStream` rather than a `FileBytes`.  The `OmfLoader` is a built-in example:

https://github.com/NationalSecurityAgency/ghidra/blob/39f388f8969198567ffda2acc3cb4273e4d7938e/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/OmfLoader.java#L131-L146

https://github.com/NationalSecurityAgency/ghidra/blob/39f388f8969198567ffda2acc3cb4273e4d7938e/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/OmfLoader.java#L378-L410

**Reply by ryanmkurtz** [reply]
So if you just have a `byte[]`, you can make a new `ByteArrayInputStream` from it and pass that in.

**Reply by ryanmkurtz** [reply]
`FileBytes` is a relatively new concept.  All loaders used to use the `InputStream` variant.

**Reply by hippietrail** [reply]
Thanks. I can't seem to spot anywhere in `OmfLoader` where the uninitialized block is getting populated. My lack of familiarity with Java surely prevents me from seeing the difference between Ghidra-specific types and functions and ones part of Java. I don't see `ByteArrayInputStream` in `OmfLoader` but that helped me find this code in `MotorolaHexLoader`
https://github.com/NationalSecurityAgency/ghidra/blob/39f388f8969198567ffda2acc3cb4273e4d7938e/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/MotorolaHexLoader.java#L326-L335
Which is pretty similar to what I'm working on. By the way, how are you quoting ranges of lines of code from the repo like that?

**Reply by ryanmkurtz** [reply]
> I can't seem to spot anywhere in `OmfLoader` where the uninitialized block is getting populated.

Uninitialized blocks, by definition, don't get populated.

> By the way, how are you quoting ranges of lines of code from the repo like that?

You can shift+click on a range of lines (in the margin) in github's presentation of the source, and then do "Copy permalink":

![image](https://github.com/user-attachments/assets/ed5bd0f4-7b5e-4fea-ab65-27b536911860)
